docs/man/powervar_cx_usb.8
'\" t
.\"     Title: powervar_cx_usb
.\"    Author: [see the "AUTHORS" section]
.\" Generator: DocBook XSL Stylesheets vsnapshot <http://docbook.sf.net/>
.\"      Date: 08/08/2025
.\"    Manual: NUT Manual
.\"    Source: Network UPS Tools 2.8.4
.\"  Language: English
.\"
.TH "POWERVAR_CX_USB" "8" "08/08/2025" "Network UPS Tools 2\&.8\&.4" "NUT Manual"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
powervar_cx_usb \- Driver for Powervar GTS and UPM Series UPS equipment with USB connection
.SH "SYNOPSIS"
.sp
\fBpowervar_cx_usb\fR \-h
.sp
\fBpowervar_cx_usb\fR \-a \fIUPS_NAME\fR [\fIOPTIONS\fR]
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
.sp
This man page only documents the hardware\-specific features of the Powervar driver\&. For information about the core driver, see \fBnutupsdrv\fR(8)\&.
.sp .5v
.RE
.SH "SUPPORTED HARDWARE"
.sp
This driver supports two Powervar UPS families with USB connections:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
UPM (either USB or Serial)
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
GTS (USB only)
.RE
.sp
For serial connected UPM Series devices, see also \fBpowervar_cx_ser\fR(8)\&.
.SH "EXTRA ARGUMENTS"
.sp
This driver supports the following optional settings in the \fBups.conf\fR(5) file:
.PP
\fBport =\fR \fIstring\fR
.RS 4
Required option for all NUT drivers\&. Some
\fIvalue\fR
must be set, typically
\fBauto\fR
for drivers that handle USB connections\&.
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
This could be a device filesystem path like
/dev/usb/hiddev0
but current use of libusb API precludes knowing and matching by such identifiers\&. They may also be inherently unreliable (dependent on re\-plugging and enumeration order)\&. At this time the actual
\fIvalue\fR
is ignored, but syntactically some
\fIport\fR
configuration must still be there\&.
.sp .5v
.RE
.RE
.sp
It is possible to control multiple UPS units simultaneously by running several instances of this driver, provided they can be uniquely distinguished by setting some combination of the \fBvendor\fR, \fBproduct\fR, \fBvendorid\fR, \fBproductid\fR, \fBserial\fR, \fBbus\fR and/or \fBdevice\fR options detailed below\&. For devices or operating systems that do not provide sufficient information, the \fBallow_duplicates\fR option can be of use (limited and risky!)
.PP
\fBvendorid =\fR \fIregex\fR, \fBproductid =\fR \fIregex\fR, \fBvendor =\fR \fIregex\fR, \fBproduct =\fR \fIregex\fR, \fBserial =\fR \fIregex\fR
.RS 4
Select a specific UPS, in case there is more than one connected via USB\&. Each option specifies an extended regular expression (see
\fBregex\fR(7)
for more information on regular expressions), which must match the UPS\(cqs entire respective
vendor/product/serial
string values (minus any surrounding whitespace), or the whole 4\-digit hexadecimal code for
vendorid
and
productid\&.
.sp
Try
\fBlsusb\fR(8)
or running this NUT driver with
\-DD
command\-line argument for finding out the strings to match\&.
.sp
Examples:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\-x vendor="Foo\&.Corporation\&.*"
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\-x vendorid="051d*"
(APC)
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
\-x product="\&.*(Smart|Back)\-?UPS\&.*"
.RE
.RE
.PP
\fBbus =\fR \fIregex\fR
.RS 4
OPTIONAL, NOT RECOMMENDED\&.
.sp
Select a UPS on a specific USB bus or group of buses\&. The argument is a regular expression that must match the bus name where the UPS is connected (e\&.g\&.
bus="002"
or
bus="00[2\-3]") as seen on Linux in
/sys/bus/usb/devices
or
\fBlsusb\fR(8); including leading zeroes\&.
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
Bus numbers are not guaranteed by the OS to be stable across re\-boots, kernel driver reloads or device re\-plugging (e\&.g\&. changing visible population of USB hubs)\&.
.sp .5v
.RE
.RE
.PP
\fBdevice =\fR \fIregex\fR
.RS 4
OPTIONAL, NOT RECOMMENDED\&.
.sp
Select a UPS on a specific USB device or group of devices\&. The argument is a regular expression that must match the device name where the UPS is connected (e\&.g\&.
device="001"
or
device="00[1\-2]") as seen on Linux in
/sys/bus/usb/devices
or
\fBlsusb\fR(8); including leading zeroes\&.
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
Device numbers are not guaranteed by the OS to be stable across re\-boots or device re\-plugging\&.
.sp .5v
.RE
.RE
.PP
\fBbusport =\fR \fIregex\fR
.RS 4
OPTIONAL, NOT RECOMMENDED\&.
.sp
If supported by the hardware, OS and libusb on the particular deployment, this option should allow to specify physical port numbers on an USB hub, rather than logical
device
enumeration values, and in turn \(em this should be less volatile across reboots or re\-plugging\&. The value may be seen in the USB topology output of
lsusb \-tv
on systems with that tool, for example\&.
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBNote\fR
.ps -1
.br
This option is not practically supported by some NUT builds (it should be ignored with a warning then), and not by all systems that NUT can run on\&.
.sp .5v
.RE
.RE
.PP
\fBallow_duplicates\fR
.RS 4
OPTIONAL, NOT RECOMMENDED\&.
.sp
If you have several UPS devices which may not be uniquely identified by the options above (e\&.g\&. only
\fIVID:PID\fR
can be discovered there), this flag allows each driver instance where it is set to take the first match if available, or proceed to try another\&.
.sp
Normally the driver initialization would abort at this point claiming "Resource busy" or similar error, assuming that the otherwise properly matched device is unique \(em and some other process already handles it\&.
.if n \{\
.sp
.\}
.RS 4
.it 1 an-trap
.nr an-no-space-flag 1
.nr an-break-flag 1
.br
.ps +1
\fBWarning\fR
.ps -1
.br
This feature is inherently non\-deterministic! The association of driver instance name to actual device may vary between runs!
.sp
If you only care to know that
\fBat least\fR
one of your no\-name UPSes is online, this option can help\&.
.sp
If you must really know
\fBwhich\fR
one, it will not!
.sp .5v
.RE
.RE
.PP
\fBusb_set_altinterface =\fR \fIbAlternateSetting\fR
.RS 4
Force redundant call to
usb_set_altinterface(), especially if needed for devices serving multiple USB roles where the UPS is not represented by the interface number
0
(default)\&.
.RE
.PP
\fBusb_config_index\fR, \fBusb_hid_rep_index\fR, \fBusb_hid_desc_index\fR, \fBusb_hid_ep_in\fR, \fBusb_hid_ep_out\fR
.RS 4
Force use of specific interface, endpoint, descriptor index etc\&. numbers, rather than defaulting to
\fI0\fR
(rarely other values in certain drivers for some devices known to use non\-zero numbers)\&. Specified as a hexadecimal number\&.
.sp
As a rule of thumb for
usb_hid_desc_index
discovery, you can see larger
wDescriptorLength
values (roughly 600+ bytes) in reports of
lsusb
or similar tools\&.
.RE
.PP
\fBLIBUSB_DEBUG =\fR \fIINTEGER\fR
.RS 4
Run\-time troubleshooting of USB\-capable NUT drivers can involve not only raising the common NUT debug verbosity (e\&.g\&. using the
DEBUG_MIN
setting in
\fBups.conf\fR(5)
or protocol commands to change the
driver\&.debug
value), but may also benefit from LibUSB specific debugging\&.
.sp
For the latter, you can set the
LIBUSB_DEBUG
driver option; alternatively you can classically export the environment variable
LIBUSB_DEBUG
before starting a NUT driver program (may be set and "exported" in driver init script or service method, perhaps via
\fBnut.conf\fR(5)), to a numeric value such as
4
("All messages are emitted")\&.
.sp
For more details, including the currently supported values for your version of the library, see e\&.g\&.:
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
https://libusb\&.sourceforge\&.io/api\-1\&.0/
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
https://libusb\&.sourceforge\&.io/api\-1\&.0/group__libusb__lib\&.html
.RE
.RE
.PP
\fBstartdelay\fR=\fInum\fR
.RS 4
Change startup delay time from 1 second default\&.
.RE
.PP
\fBoffdelay\fR=\fInum\fR
.RS 4
Change shutdown delay time from 0 second default\&.
.RE
.PP
\fBdisptesttime\fR=\fInum\fR
.RS 4
Change display test time from the 10 second default (UPM only)\&.
.RE
.SH "INSTANT COMMANDS"
.sp
This driver supports the following Instant Commands (see \fBupscmd\fR(8)):
.SS "All GTS and UPM UPS units"
.PP
\fBbeeper\&.disable\fR
.RS 4
Disable the UPS beeper\&.
.RE
.PP
\fBbeeper\&.enable\fR
.RS 4
Enable the UPS beeper\&.
.RE
.PP
\fBbeeper\&.mute\fR
.RS 4
Temporarily mute the UPS beeper\&.
.RE
.PP
\fBshutdown\&.return\fR
.RS 4
Turn off the load possibly after a delay and return when power is back\&.
.RE
.PP
\fBshutdown\&.stop\fR
.RS 4
Stop a shutdown in progress\&.
.RE
.PP
\fBshutdown\&.reboot\fR
.RS 4
Shut down the load briefly while rebooting the UPS\&.
.RE
.PP
\fBshutdown\&.stayoff\fR
.RS 4
Turn off the load and remain off\&.
.RE
.PP
\fBtest\&.battery\&.start\&.deep\fR
.RS 4
Start a "deep" battery test\&. This test runs the UPS until the low battery point and then returns to the AC line\&.
.RE
.PP
\fBtest\&.battery\&.start\&.quick\fR
.RS 4
Start a "quick" battery test\&. The default time is 2 minutes\&. This time can be set in the
\fBups\&.conf\fR
file\&. See
\fBtestime\fR
above\&.
.RE
.PP
\fBtest\&.battery\&.stop\fR
.RS 4
Stops a battery test that is in progress\&.
.RE
.PP
\fBtest\&.failure\&.start\fR
.RS 4
Starts a simulation of an input power failure\&.
.RE
.PP
\fBtest\&.failure\&.stop\fR
.RS 4
Stops a simulation of an input power failure\&.
.RE
.SS "All UPM UPS units"
.PP
\fBtest\&.panel\&.start\fR
.RS 4
Start testing the UPS panel\&.
.RE
.PP
\fBreset\&.input\&.minmax\fR
.RS 4
Reset minimum and maximum input voltage status\&.
.RE
.SH "WRITABLE VARIABLES"
.sp
See \fBupsrw\fR(8) to see what variables are writable for the UPS\&.
.SH "AUTHORS"
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
Bill Elliot <bill@wreassoc\&.com>
.RE
.SH "SEE ALSO"
.sp
For serial connected UPM Series devices, see \fBpowervar_cx_ser\fR(8)\&.
.SS "The core driver:"
.sp
\fBnutupsdrv\fR(8)
.SS "Internet resources:"
.sp
The NUT (Network UPS Tools) home page: https://www\&.networkupstools\&.org/historic/v2\&.8\&.4/


NEWS
Please see NEWS.adoc for actual contents


drivers/powervar_cx_usb.c
/*vim ts=4*/
/* powervar_cx_usb.c - USB Driver for Powervar UPSs using CUSPP.
 *
 * Supported Powervar UPS families in this driver:
 * UPM Family (All)
 * GTS Family (All)
 *
 * Copyright (C)
 *     2024, 2025 by Bill Elliot <bill@wreassoc.com>
 *     (USB comms based on tripplite_usb.c and other drivers)
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 * History:
 * - 7 February 2025, Bill Elliot
 * Working well with USB (or serial for -cs driver)
 * - 2 December 2024.  Bill Elliot
 * Started to work on common variable and function file so this driver and the
 *  serial (-cs) driver can make use of the same protocol related functions.
 * - 2 October 2024.  Bill Elliot
 * Used pieces of oneac.c driver to get jump-started.
 *
 */

#include "main.h"
#include "nut_stdint.h"
#include "nut_libusb.h"

#include "usb-common.h"

/* Two drivers include the following. Provide some identifier for any differences */
#define PVAR_USB	1	/* This is the USB comm driver */
#include "powervar_cx.h"	/* Common driver variables and functions */

#define DRIVER_NAME	"Powervar-CUSSP UPS driver (USB)"
#define DRIVER_VERSION	"1.00"

/* USB comm stuff here */
#define USB_RESPONSE_SIZE	8
#define MAX_CNCT_ATTMPTS	60	/* x calls to upsdrv_updateinfo */

/* Powervar */
#define POWERVAR_VENDORID	0x4234

#define BIG_BUFFER		512

/* Flag indicating a disconnected USB cable...and need to reconnect */
static unsigned int ReconnectFlag = 0;

static USBDevice_t *hd = NULL;
static USBDevice_t curDevice;
static USBDeviceMatcher_t *reopen_matcher = NULL;
static USBDeviceMatcher_t *regex_matcher = NULL;
static usb_dev_handle *udev;
static usb_communication_subdriver_t *comm_driver = &usb_subdriver;

/* USB IDs device table */
static usb_device_id_t powervar_usb_device_table[] = {
	/* GTS ... */
	{ USB_DEVICE(POWERVAR_VENDORID, 0x0001), NULL },

	/* UPM ... */
	{ USB_DEVICE(POWERVAR_VENDORID, 0x0002), NULL },

	/* Terminating entry */
	{ 0, 0, NULL }
};


/* driver description structure */
upsdrv_info_t upsdrv_info = {
	DRIVER_NAME,
	DRIVER_VERSION,
	"Bill Elliot <bill@wreassoc.com>",
	DRV_BETA,
	{ NULL }
};


static int subdriver_match_func(USBDevice_t *arghd, void *privdata)
{
	NUT_UNUSED_VARIABLE(privdata);

	/* FIXME? Should we save "arghd" into global "hd" variable?
	 * This was previously shadowed by function argument named "hd"...
	 */
	/* hd = arghd; */

	printf ("In subdriver_match_func...");

	switch (is_usb_device_supported(powervar_usb_device_table, arghd))
	{
		case SUPPORTED:
			printf ("SUPPORTED\n");
			return 1;

		case POSSIBLY_SUPPORTED:
			printf ("POSSIBLY_SUPPORTED\n");
			return 0;

		case NOT_SUPPORTED:
			printf ("NOT_SUPPORTED\n");
			return 0;
		default:
			return 0;
	}
}


static USBDeviceMatcher_t subdriver_matcher = {
	&subdriver_match_func,
	NULL,
	NULL
};


static int match_by_something(usb_dev_handle *argudev, USBDevice_t *arghd, usb_ctrl_charbuf rdbuf, usb_ctrl_charbufsize rdlen)
{
	int i;
	char sData[BIG_BUFFER];
	size_t sDataLen = BIG_BUFFER;

	NUT_UNUSED_VARIABLE(argudev);
	NUT_UNUSED_VARIABLE(arghd);
	NUT_UNUSED_VARIABLE(rdbuf);
	NUT_UNUSED_VARIABLE(rdlen);

	GetUPSData (PID_REQ, sData, sDataLen);

	for (i = 0; i < BIG_BUFFER ; i++)
	{
		if (sData[i] == ENDCHAR)
		{
			break;
		}
	}

	/* Just trying to be nice to the terminal output by keeping the CR out of string... */
	sData[i] = 0;
	upsdebugx(3, "'GetUPSData' returned '%s', i: %d", sData, i);
	sData[i] = ENDCHAR;	/* ... until here. */

	sData[i+1] = 0;		/* Force trailing null */

	if (GetSubstringPosition ((const char*)sData, PID_PROT_DATA) == 1)
	{
		upsdebugx(3, "Powervar CUSPP device found!");
	}

	return 1;
}


/***********************************
 * USB communication functions     *
 **********************************/

/* This function is called to flush the USB hardware of any remaining data */
/* Would like to find a way to avoid 'nut_libusb_get_interrupt: Connection timed out' */
/*  debug level 2 message. */
static void USBFlushReceive (void)
{
	unsigned int uiCount = 0;
	char response_in[USB_RESPONSE_SIZE + 1];

	upsdebugx(3, "Flushing USB receive.");

	while (comm_driver->get_interrupt(udev,
			(usb_ctrl_charbuf)response_in,
			(usb_ctrl_charbufsize) USB_RESPONSE_SIZE,
			(usb_ctrl_timeout_msec) 100) > 0)
	{
		uiCount++;
	}

	upsdebugx(6, "Flush Count: %u.", uiCount);
}


/* This function is called to send the data request to the initialized device. */
/* Return value should be >0 for success. */
size_t SendRequest (const char* sRequest)
{
	char outbuff[40];
	int ret = 0;
	size_t ReqLen = strlen(sRequest);
	size_t i;

	/* Last char reserved for ENDCHAR, so sizeof-1 for the string */
	if (sizeof(outbuff) - 1 < ReqLen) {
		upsdebugx(1, "SendRequest(): request too long, will be truncated");
		/* FIXME(no): unexpected situation, abort? */
		/* Since problem is logged above and we can not abort for all
		 *  calls to this function, we have nothing more to do. An
		 *  improperly formatted request will be ignored by the unit.
		 */
	}

	/* Clear output buffer area */
	memset(outbuff, 0, sizeof(outbuff));

	/* Move command into USB buffer and add terminating character
	 * Not strncpy because we want to add ENDCHAR to a known position anyway
	 */
	for(i = 0; i < ReqLen && i < sizeof(outbuff) - 1; i++)
	{
		outbuff[i] = sRequest[i];
	}

	upsdebugx(3, "SendRequest('%s', size: %" PRIuSIZE ")", outbuff, ReqLen);

	outbuff[i] = ENDCHAR;
	ReqLen++;			/* Add one for added CR */

	ret = comm_driver->set_report(udev, 0,
		(usb_ctrl_charbuf)outbuff,
		(usb_ctrl_charbufsize)ReqLen);

	upsdebugx(5, "set_report ret:   %d", ret);

	if (ret < 0 || (size_t)ret != ReqLen)
	{
		upsdebugx(1, "libusb_set_report() returned %d instead of %" PRIuSIZE, ret, ReqLen);
	}

	if (ret < 1)
		return 0;

	return (size_t)ret;
}

/* Get the response from the UPS */
/* chBuff is the buffer to receive the response into */
/* BuffSize is the size, in chars, of chBuff */
ssize_t PowervarGetResponse (char* chBuff, const size_t BuffSize)
{
	unsigned char response_in[USB_RESPONSE_SIZE + 2];
	int ret = 0, done = 0;
	size_t i = 0;
	size_t j = 0;

	ssize_t return_val = 1;		/* Set up for a bad return */

	memset(response_in, 0, sizeof(response_in));

	for (i = 0 ; (done == 0) && (i < BuffSize) ; )
	{
		upsdebugx(5, "Receive Loop: %" PRIuSIZE, i);

		ret = comm_driver->get_interrupt(udev,
			(usb_ctrl_charbuf)response_in,
			(usb_ctrl_charbufsize) USB_RESPONSE_SIZE,
			(usb_ctrl_timeout_msec) 500);

		if (ret > 0)
		{
			for (j=0 ; j < USB_RESPONSE_SIZE ; j++)
			{
				if (response_in[j] == ENDCHAR)
				{
					upsdebugx(5, "<CR> found in response @ pos: %" PRIuSIZE ".", j);
					done = 1;
					break;
				}

				chBuff[i] = response_in[j];

				i++;
			}

			upsdebugx(5, "Loop Response: '%s'", response_in);

			/* Clear response buffer for next chunk */
			memset(response_in, 0, sizeof(response_in));
		}
		else
		{
			upsdebugx(1, "Unexpected return value: %d", ret);
			break;
		}
	}

	upsdebugx (4, "PowervarGetResponse buffer: '%s'", chBuff);

	if (ret > 0)
	{
		USBFlushReceive ();	/* Clear USB hardware */
		return_val = 0;		/* Comms OK */
	}
	else
	{
		ReconnectFlag = 1;	/* Attempt reconnection */
	}

	return return_val;
}


/****************************************************************
 * Below are the primary commands that are called by main       *
 ***************************************************************/

void upsdrv_initups(void)
{
	char *regex_array[USBMATCHER_REGEXP_ARRAY_LIMIT];
	int ret = 1;

	upsdebugx(2, "In upsdrv_initups");

	warn_if_bad_usb_port_filename(device_path);

	/* Process the UPS selection options */
	regex_array[0] = NULL; /* handled by USB IDs device table */
	regex_array[1] = getval("productid");
	regex_array[2] = getval("vendor");	/* vendor string */
	regex_array[3] = getval("product"); 	/* product string */
	regex_array[4] = getval("serial");
	regex_array[5] = getval("bus");
	regex_array[6] = getval("device");
#if (defined WITH_USB_BUSPORT) && (WITH_USB_BUSPORT)
	regex_array[7] = getval("busport");
# else
	if (getval("busport")) {
		upslogx(LOG_WARNING, "\"busport\" is configured for the device, but is not actually handled by current build combination of NUT and libusb (ignored)");
	}
# endif

	ret = USBNewRegexMatcher(&regex_matcher, regex_array, REG_ICASE | REG_EXTENDED);
	if (ret == -1)
	{
		fatal_with_errno(EXIT_FAILURE, "USBNewRegexMatcher");
	}
	else if (ret)
	{
		fatalx(EXIT_FAILURE, "invalid regular expression: %s", regex_array[ret]);
	}

	/* link the matchers */
	regex_matcher->next = &subdriver_matcher;

	upsdebugx(2, "Trying to open USB device...");

	/* Search for the first supported UPS matching the regular expression */
	ret = comm_driver->open_dev(&udev, &curDevice, regex_matcher, match_by_something);
	if (ret < 1)
	{
		fatalx(EXIT_FAILURE, "No matching USB/HID UPS found");
	}

	hd = &curDevice;

	upsdebugx(1, "Detected a UPS: %s/%s", hd->Vendor ? hd->Vendor : "unknown", hd->Product ? hd->Product : "unknown");

	dstate_setinfo("ups.vendorid", "%04x", hd->VendorID);
	dstate_setinfo("ups.productid", "%04x", hd->ProductID);

	upsdebugx(3, "VendorID: %04x", hd->VendorID);
	upsdebugx(3, "ProductID: %04x", hd->ProductID);

	/* create a new matcher for later reopening */
	ret = USBNewExactMatcher(&reopen_matcher, hd);
	if (ret)
	{
		fatal_with_errno(EXIT_FAILURE, "USBNewExactMatcher");
	}

	/* link the two matchers */
	reopen_matcher->next = regex_matcher;
}

void upsdrv_initinfo(void)
{
	upsdebugx(3, "In upsdrv_initinfo");

	/* Get port ready */
	SendRequest ((const char*)COMMAND_END);		/* Just get device ready -- flush */
	USBFlushReceive ();				/* Just flush response */

	PvarCommon_Initinfo ();
}


/* This function is called regularly for data updates. */
void upsdrv_updateinfo(void)
{
	int ret;
	static int CnctAttempts = 0;

	upsdebugx(3, "In upsdrv_updateinfo");

	if (ReconnectFlag)
	{
		dstate_setinfo("driver.state", "reconnect.trying");
		upslogx(LOG_WARNING, "USB device may be detached.");
		upslogx(LOG_NOTICE, "USB reconnect attempt: %d.", ++CnctAttempts);
		upsdebugx(4, "USB reconnect attempt: %d", CnctAttempts);

		hd = NULL;

		ret = comm_driver->open_dev(&udev, &curDevice, reopen_matcher, match_by_something);
		if (ret < 1)
		{
			if (CnctAttempts >= MAX_CNCT_ATTMPTS)
			{
				upsdebugx(4, "Exceeded max reconnect attemtps.");
				fatalx(EXIT_FAILURE, "Exceeded max reconnect attempts.");
			}
			else
			{
				upslogx(LOG_INFO, "USB reconnect attempt %d failed.", CnctAttempts);
				upslogx(LOG_INFO, "Will try another reconnect in a bit.");
				dstate_datastale();
				return;
			}
		}

		upsdebugx(4, "USB device reconnected!");
		ReconnectFlag = 0;	/* Show good for now */
		CnctAttempts = 0;

		hd = &curDevice;

		upslogx(LOG_NOTICE, "USB reconnect successful");
		dstate_setinfo("driver.state", "reconnect.updateinfo");
		upsdrv_initinfo();

		dstate_setinfo("driver.state", "quiet");
	}

	PvarCommon_Updateinfo ();
}


/**************************************
 * Handlers for NUT command calls     *
 *************************************/

void upsdrv_help(void)
{
	printf("\n---------\nNOTE:\n");
	printf("This driver is for connecting to a Powervar UPS USB port.\n");
}

void upsdrv_cleanup(void)
{
	upsdebugx(2, "In upsdrv_cleanup");

	comm_driver->close_dev(udev);
	USBFreeExactMatcher(reopen_matcher);
	USBFreeRegexMatcher(regex_matcher);
	free(curDevice.Vendor);
	free(curDevice.Product);
	free(curDevice.Serial);
	free(curDevice.Bus);
	free(curDevice.Device);
#if (defined WITH_USB_BUSPORT) && (WITH_USB_BUSPORT)
	free(curDevice.BusPort);
#endif
}

/* End of powervar_cx_usb.c file */
